nitro-protocol/contracts/LedgerFinancingApp.sol

```diff
@@ -0,0 +1,151 @@
+pragma solidity 0.8.17;
+pragma experimental ABIEncoderV2;
+
+import './interfaces/IForceMoveApp.sol';
+import './libraries/NitroUtils.sol';
+import {ExitFormat as Outcome} from '@statechannels/exit-format/contracts/ExitFormat.sol';
+
+// LedgerFinancingApp is a ForceMoveApp that allows a intermediary to earn interest
+// on a deposit. It fuctions as a ConsensusApp with the following additional rule:
+//   - the intermediary can unilaterally transition from state n to state n+1,
+//     forcing calculated interest into the intermediary's Outcome allocation
+contract LedgerFinancingApp is IForceMoveApp {
+    struct Funds {
+        address[] asset;
+        uint256[] amount;
+    }
+
+    struct InterestAppData {
+        // a per-day simple interest rate (daily percentage yield), as a fraction.
+        // ie, 1% per day is represented with dpyNum = 1 and dpyDen = 100.
+        //
+        // Lower numbers (eg, fraction in simplest terms) produce least risk of overflow.
+        uint128 dpyNum;
+        uint128 dpyDen;
+        // the block number of the latest principal adjustment
+        uint256 blocknumber;
+        // the current principal. Decreases as the serviceProvider earns via the channel.
+        Funds principal;
+        // the total interest collected so far. Strictly increasing.
+        // The value of the intermediary's allocation can never be less than this:
+        // ie, when the intermediary's collectedInterest grows to be equal to their
+        // allocation, the channel is effectively spent and can be concluded.
+        Funds collectedInterest;
```

Comment:
I don't believe this is actually used?

Reply:
This is interesting. No, this is not used in `requireStateSupported`, and so it could, in practical terms, be dropped from the appData. But it's instrumental in the correct operation of the channel, since it does the accounting on the intermediary's profits. When the Intermediary's allocation in the channel falls down to be equal to this amount, she has no future interest in the channel, so it should be closed.

If not in the appData, this information would have to be stored in a separate structure somewhere else (appMetaData). We have no infrastructure for storing runtime metadata for a specific channel.

Should we take this as a hint that we might want that?

Reply:
I'm a bit confused by this. Why does the intermediary's allocation matter?

Let's say the principal is 100, and after some time the hub is due 3 coins of interest. Meanwhile, the provider has earned 97 coins so that the hub has 3 coins left. It seems to me that the hub does still have an interest in the channel, since if nothing else happens (just time lapses), he will start to earn more interest which should be claimable (the service provider has plenty of cash to pay the interest).